src/depends/libDatabase/LevelDB.cpp

```diff
@@ -15,610 +15,493 @@
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
  */
 
-#include <string>
-
-
 #include "LevelDB.h"
 #include "common/Constants.h"
-#include "depends/common/Common.h"
 #include "depends/common/CommonData.h"
-#include "depends/common/FixedHash.h"
 #include "libUtils/Logger.h"
 
 using namespace std;
 
-void LevelDB::log_error(leveldb::Status status) const
-{
-    if(!status.IsNotFound())
-    {
-     LOG_GENERAL(WARNING, "LevelDB " << m_dbName << " status is not OK - " << status.ToString());
-    }
-}
+namespace {
 
-LevelDB::LevelDB(const string& dbName, const string& path, const string& subdirectory)
-{
-    this->m_subdirectory = subdirectory;
-    this->m_dbName = dbName;
-    this->m_db = NULL;
+template <typename ArgT>
+std::string LookupImpl(const std::shared_ptr<leveldb::DB> db,
+                       const std::string& dbName, ArgT&& arg,
+                       bool* found = nullptr) {
+  std::string value;
+  if (!db) {
+    LOG_GENERAL(WARNING, "LevelDB " << dbName << " isn't available");
+    if (found) {
+      *found = false;
+    }
```

Comment:
@yaron-zilliqa Sorry, I'm not understanding this properly, but the new code does jump out to me as a bit scary. I agree that its dangerous to assume 'DB doesn't exist' is the same as 'DB exists, but is empty'.

Reply:
The methods haven't been reordered. The only change in logic here is checking for a nullptr before accessing it.
The previous methods were really a copy of one another with a different construction of leveldb::Slice, so the not simply invoke XXXXXImpl(...) instead of repeating the code.
The problem it solves is that the process won't crash (and then redownload persistence upon restart) if the required LevelDB object was NULL. I came across this scenario this time round when setting ARCHIVAL_LOOKUP_WITH_TX_TRACES to true and calling ots_getTransactionError.